From 789e65361055c9786f8548aa1c7f022dbf99b21f Mon Sep 17 00:00:00 2001
From: Matthew Lugg <mlugg@mlugg.co.uk>
Date: Wed, 19 Nov 2025 17:08:49 +0000
Subject: [PATCH 067/133] std.http: stop assuming previous chunk state
X-Developer-Signature: v=1; a=openpgp-sha256; l=1549; i=xtex@astrafall.org;
 h=from:subject; bh=40QVm81GvM6L9IEZlQDxg79XU67QVZzau1eeApegqPM=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78limSqQ0qjWkPt74WN7zg6SqAW/V7GsFL3VCJb0XL
 371dr5yx0QWBjEuBksxRZYiwwZvVp10ftFl5bIwc1iZQIZIizQwMDAwsDDw5SbmlRrpGOmZahvq
 GRrqGOkYMXBxCsBUn/jLyLC9/XnvRtXz6TpGR62UDKsMVAu+7rHSf6LwIPqUts5lkVUM/zS95+1
 YV/T3/0RrVkOu8HdL3xz9eoNFwLV22tYbXw+1HecEAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

The full file may not be written, either due to a previous chunk being
in-progress when `sendFile` was called, or due to `limit`.

(cherry picked from commit b05fefb9c9bf3065e353102d0e54ea76d5f4e34d)
---
 lib/std/http.zig | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/lib/std/http.zig b/lib/std/http.zig
index dcc96ba741ce..eedd729576e0 100644
--- a/lib/std/http.zig
+++ b/lib/std/http.zig
@@ -962,6 +962,7 @@ pub const BodyWriter = struct {
             // have to flush the chunk header before knowing the chunk length.
             return error.Unimplemented;
         };
+        if (data_len == 0) return error.EndOfStream;
         const out = bw.http_protocol_output;
         l: switch (bw.state.chunk_len) {
             0 => {
@@ -975,8 +976,7 @@ pub const BodyWriter = struct {
             2 => {
                 try out.writeAll("\r\n");
                 bw.state.chunk_len = 0;
-                assert(file_reader.atEnd());
-                return error.EndOfStream;
+                continue :l 0;
             },
             else => {
                 const chunk_limit: std.Io.Limit = .limited(bw.state.chunk_len - 2);
@@ -985,8 +985,7 @@ pub const BodyWriter = struct {
                 else
                     try out.write(chunk_limit.slice(w.buffered()));
                 bw.state.chunk_len -= n;
-                const ret = w.consume(n);
-                return ret;
+                return w.consume(n);
             },
         }
     }
-- 
2.52.0

